src/stores/index.ts

```diff
@@ -20,16 +35,316 @@ export const getState: (key: keyof Store) => any = key => {
+    interfaceId: '',
+    lsp2Schema: null,
+  },
+}
+
+export type TokenInfo = {
+  type: LSPType
+  address?: string
+  name?: string
+  symbol?: string
+  decimals?: string
+  balance?: number
+  label?: string
+}
+
+const detectLSP = async (
+  contractAddress: string,
+  lspType: Exclude<LSPType, LSPType.Unknown>,
+  owned = false
+): Promise<TokenInfo | undefined> => {
+  if (
+    lspType in
+    {
+      [LSPType.Unknown]: true,
+      [LSPType.EoA]: true,
+      [LSPType.SC]: true,
+      [LSPType.UP]: true,
+    }
+  ) {
+    return undefined
+  }
+
+  const { contract: Contract } = useWeb3()
+  // EIP-165 detection
+  const contract = Contract(eip165ABI.concat(erc20ABI) as any, contractAddress)
+
+  // Check if the contract implements the LSP interface ID
+  let doesSupportInterface: boolean
+  try {
+    doesSupportInterface = await contract.methods
+      .supportsInterface(lspTypeOptions[lspType].interfaceId)
+      .call()
+  } catch (error) {
+    doesSupportInterface = false
+  }
+  if (!doesSupportInterface) {
+    return undefined
+  }
+
+  try {
+    let currentDecimals = '0'
+    let balance = owned ? 1 : 0
+    try {
+      currentDecimals = await contract.methods.decimals().call()
+
+      if (currentDecimals !== '0') {
+        const _balance = await contract.methods
+          .balanceOf(store['address'])
+          .call()
+          .catch(() => undefined)
+        balance = _balance
+          ? new BN(_balance, 10)
+              .div(new BN(10).pow(new BN(currentDecimals || '0', 10)))
+              .toNumber()
+          : 0
+      }
+    } catch (err) {
+      console.error(contractAddress, lspType, err, 'no balance')
+    }
+    // ERC725 detection
+    const { getInstance } = useErc725()
+
+    const erc725 = await getInstance(contractAddress)
+    const [{ value: name }, { value: symbol }] = await erc725.fetchData([
+      'LSP4TokenName',
+      'LSP4TokenSymbol',
+    ])
+    if (typeof name !== 'string' || typeof symbol !== 'string') {
+      throw new Error('Unable to get name and/or symbol')
+    }
+    let shortType: string = lspType
+    switch (shortType) {
+      case LSPType.LSP7DigitalAsset:
+        shortType = 'LSP7'
+        break
+      case LSPType.LSP8IdentifiableDigitalAsset:
+        shortType = 'LSP8'
+        break
+      case LSPType.LSP3UniversalProfileMetadata:
+        shortType = 'LSP3'
+        break
+      case LSPType.LSP9Vault:
+        shortType = 'LSP9'
+        break
+    }
+    console.log('success', contractAddress, lspType)
+    return {
+      type: lspType,
+      name,
+      symbol,
+      address: contractAddress,
+      balance,
+      decimals: currentDecimals,
+      label: `${shortType} ${name} (${symbol}) ${contractAddress.substring(
+        0,
+        10
+      )}...`,
+    }
+  } catch (err) {
+    console.error(contractAddress, lspType, err)
+    return undefined
+  }
+}
+
+export const sampleEoA = '0x311611C9A46a192C14Ea993159a0498EDE5578aC'
```

Comment:
Maybe let's create separate `constants` file, I wouldn't put everything into store file, it should really just keep app state nothing else..

Reply:
Sure